### PR TITLE
Add backend oauth

### PR DESCRIPTION
**Pull request by nbarnabee**
**UPDATE:**  This request is ready for review!

Things are slowly coming along.  This is still very much a WIP.

The data I was trying to pass wasn't making it past the test server's validation, but I had to pass the authentication check to get that far, and that's all I really cared about.

My next step will be to write a route and associated models/schemas to handle PUT requests to `/api/tools/{tool}/annotations` and see if I can successfully update one of the test tools I've made.  Once I'm able to do that, I will consider this pull request ready for review.

### CODE REVIEW

**Comment by nbarnabee** [comment]
https://toolhub-demo.wmcloud.org/api/auditlogs/6878/

I greatly simplified my sample test and was able to confirm that the connection is working.  

Is this enough for an initial pull request?  (I am trying to be more mindful about the amount of changes I shove into these things.)

File: app.py [hunk 1]
```diff
@@ -2,7 +2,86 @@
...
+from authlib.integrations.flask_client import OAuth
+
+load_dotenv()
+
+oauth = OAuth(app)
+# Note that the name of the env variables needs to match the registered name here
+# See authlib documentation for more information
+oauth.register(
+    name="toolhub",
+    access_token_url="https://toolhub-demo.wmcloud.org/o/token/",
```
Comment (blancadesal) [inline app.py:24]: You can delete all of this and leave just `name="toolhub"` here, then plug the URLs into the config file (in the places where we currently have URLs pointing to localhost:5000). I'm the one who confused you with this in the [Phab ticket ](https://phabricator.wikimedia.org/T327856) I'm afraid :/ I was testing a bunch of stuff to make things work, not all were relevant.

What happens is that anything you put inside `oauth.register()` will override whatever is in your config file; conversely, if `oauth.register()` doesn't specify anything, the oauth lib will look inside the config file and get the params from there.

File: api/config.py [hunk 2]
```diff
@@ -15,6 +18,29 @@ class BaseConfig:
...
+  SECRET_KEY = os.getenv(
+    "SECRET_KEY",
+    default=os.urandom(128).decode("utf8", errors="ignore"),
+)
+  SERVER_NAME = os.getenv("SERVER_NAME", default="localhost:5000")
+  TOOLHUB_CLIENT_ID = os.getenv("TOOLHUB_CLIENT_ID")
+  TOOLHUB_CLIENT_SECRET = os.getenv("TOOLHUB_CLIENT_SECRET")
+  TOOLHUB_ACCESS_TOKEN_URL = os.getenv(
+    "TOOLHUB_ACCESS_TOKEN_URL",
+    default="http://web:5000/o/token/",
```
Comment (blancadesal) [inline api/config.py:31]: Plug the Toolhub demo server URLs in here

File: api/config.py [hunk 3]
```diff
@@ -15,6 +18,29 @@ class BaseConfig:
...
+)
+  TOOLHUB_AUTHORIZE_URL = os.getenv(
+    "TOOLHUB_AUTHORIZE_URL",
+    default="https://localhost:5000/authorize/",
```
Comment (blancadesal) [inline api/config.py:35]: idem

File: api/config.py [hunk 4]
```diff
@@ -15,6 +18,29 @@ class BaseConfig:
...
+)
+  TOOLHUB_API_BASE_URL = os.getenv(
+    "TOOLHUB_API_BASE_URL",
+    default="http://web:5000/api/",
```
Comment (blancadesal) [inline api/config.py:39]: idem

**Review by blancadesal — CHANGES_REQUESTED** [review]
Good job figuring all of this out. I've left a few remarks regarding oauth config, but otherwise this is good to go!

The merge conflicts come from me merging your previous patch. I _think_ rebasing this PR should help vs. manually resolving conflicts.

File: app.py (hunk 1 above)
Comment (nbarnabee) [reply app.py:24]: That's what I thought would probably happen, but I also thought, "When in doubt let's just put all the things in all the places and then sort it out later."

**Comment by nbarnabee** [comment]
Force-pushed due to rebase.  Next I'll address the comments.

**Comment by nbarnabee** [comment]
[2506871](https://github.com/wikimedia/toolhunt/pull/23/commits/2506871f6ea1140787266ece562736cae8505ef3) addresses changes requested by @blancadesal

**Review by blancadesal — APPROVED** [review]

**Pull request merged**